### ISSUE TITLE
Running containers from BuildKit results without images

### ISSUE THREAD

**Issue by tonistiigi** [moby/moby#40887]
During the development of container applications, most time is spent on doing iterative code changes, rebuilding code with `docker build` and then running it with `docker run`. 

One of the performance bottlenecks in this flow is image creation. Every time you need to invoke `docker run` you need to point it to an image. Image creation is slow because images are immutable content-addressable structures defined by tarballs. So to create an image, we need to convert its snapshots to tarballs and then create an image config based on them. In moby case, we then extract the tar metadata from tarballs to disk and discard the actual duplicate data. In containerd storage, this is actually even much slower as containerd keeps the original data, and therefore all layers need to be gzip compressed as well. As a rough estimation 1 sec per 100MB is spent on waiting for this (in some cases more). If your images get larger, this problem gets more and more important. Some languages that keep a lot of source files like Nodejs are also more affected by this than others.

This all results in significant developer overhead compared to host development that can't be optimized out this a smarter Dockerfile. Although almost all of the time, these images as discarded as soon as another code change has been made.

Buildkit has already made some steps to solve this. Compared to the legacy builder, no image is created after every command, and only the very final build result is converted to an image. Eg. for multi-stage builds, no images are created for intermediate stages, already resulting in a better experience. If another output type is specified with `--output`, no image is created at all. Buildkit runs its internal containers directly from snapshots without an image config.

This ability should be exposed to the `docker build` users as well. Images should only be needed for releases and distribution. For development, we can run containers directly from the build cache. This also makes storage management much easier as build-cache can be managed with GC and intelligent `prune` commands while when we create an image, it is never released until the user runs a command to delete it.

For implementation, there are two possibilities. One is to make the build targets available to `docker run`. With this method, when a `Solve()` request with the gateway API returns, it carries an ID that can be used for `docker run <build-result-id>`. When container is created, dockerd will recognize that ID and ask a snapshot from the builder to get the rootfs. Once both the gateway is released and container is deleted, resources can be released from build cache.

Another implementation is to do this only in BuildKit side, using https://github.com/moby/buildkit/issues/749, which allows running arbitrary commands with BuildKit mounts. The benefit of this approach is that it can be used with all BuildKit configurations. A particularly interesting one is using `docker buildx` with the docker container driver. Currently, this method requires using `--load` if you want to make an image available to `docker run`, so it is even slower. The downside of using `Exec` method in BuildKit is that we will not be able to support all the flags that `docker run` has, only the very minimal ones. Usually, this should be ok for development though.

Ideally, both of the methods could be implemented, but BuildKit `exec` approach seems more critical.

Once this is done, another level of optimization would be to use it together with `MergeOp` https://github.com/moby/buildkit/issues/1431 that also would eliminate the overhead of `COPY` operations duplicating source code after every change, significantly improving working with bigger images.

**Comment by tao12345666333** [comment]
SGTM. This is great for improving development speed.

**Comment by thaJeztah** [comment]
related: https://github.com/moby/moby/issues/40379

**Comment by hinshun** [comment]
> Once this is done, another level of optimization would be to use it together with MergeOp moby/buildkit#1431 that also would eliminate the overhead of COPY operations duplicating source code after every change, significantly improving working with bigger images.

@tonistiigi To be clear, this is referring to mounting in the source in a `MergeOp` to avoid copying source files right?

**Comment by tonistiigi** [comment]
@hinshun Yes, `COPY` would invoke a merge internally and container would run directly from merge point. Once container is released, source files can be again reused for rsync destination without any need to make copies of them.
